lightning/src/blinded_path/utils.rs

```diff
@@ -170,17 +170,83 @@ fn encrypt_payload<P: Writeable>(payload: P, encrypted_tlvs_rho: [u8; 32]) -> Ve
 	write_adapter.encode()
 }
 
-/// Blinded path encrypted payloads may be padded to ensure they are equal length.
-///
-/// Reads padding to the end, ignoring what's read.
-pub(crate) struct Padding {}
+/// Represents the padding round off size (in bytes) that is used to pad [`BlindedHop`]
+pub const PADDING_ROUND_OFF: usize = 50;
```

Comment:
> Why 50? Are common blinded path hops 50 bytes?

Seems this would depend on the received TLVs' size (i.e., how big `MessageContext` or `PaymentContext` is for the given use case).

> We should at least write down how big blinded path hops generally are to make sure we're not cutting right in the middle of common sizes here.

Could you elaborate on this concern?

Reply:
First off, I was able to determine the common sizes for various message and payment TLVs. You can check out the [branch](https://github.com/shaavan/rust-lightning/commits/padding-test) where I introduced the test.

After that, I ran some experiments with different padding sizes. It looks like P=50 and P=70 are both solid options. P=50 gave us 3 unique groups, and P=70 reduced it to 2 while keeping the extra padding fairly minimal.

Here's a table for comparison:

| Original Sizes | Rounded (P=50) | Rounded (P=70) |
|----------------|----------------|----------------|
| 35             | 50             | 70             |
| 70             | 100            | 70             |
| 10             | 50             | 70             |
| 28             | 50             | 70             |
| 62             | 100            | 70             |
| 96             | 100            | 140            |
| 96             | 100            | 140            |
| 96             | 100            | 140            |
| 27             | 50             | 70             |
| 42             | 50             | 70             |
| 26             | 50             | 70             |
| 26             | 50             | 70             |
| 128            | 150            | 140            |
| 53             | 100            | 70             |

Let me know which size you think we should go with! I'm happy to make any adjustments.

Thanks so much!

Reply:
Could you label the rows by their types?

Reply:
Hi Jeff! Thank you so much for your patience! I’ve updated the table with the labels for each row, as requested. Here’s the latest version:


| Type                | Property                                    | Original Sizes | Rounded (P=50) | Rounded (P=70) |
|---------------------|---------------------------------------------|----------------|----------------|----------------|
| **Message ForwardTlvs**                                                                                                    |
|                     | With NodeId + No Blinding override          | 35             | 50             | 70             |
|                     | With NodeId + With Blinding override        | 70             | 100            | 70             |
|                     | With SCID + No Override                     | 10             | 50             | 70             |
| **Message ReceiveTlvs**                                                                                                    |
|                     | MessageContext::Offers::InvoiceRequest      | 28             | 50             | 70             |
|                     | MessageContext::Offers::OutboundPayment     | 62             | 100            | 70             |
|                     | MessageContext::Offers::OutboundPayment     | 96             | 100            | 140            |
|                     | MessageContext::Offers::InboundPayment      | 96             | 100            | 140            |
|                     | MessageContext::AsyncPayments               | 96             | 100            | 140            |
|                     | MessageContext::DNSResolver                 | 27             | 50             | 70             |
|                     | MessageContext::Custom                      | 42             | 50             | 70             |
| **Payment ForwardTlvs**                                                                                                    |
|                     | With no Blinding Override                   | 26             | 50             | 70             |
|                     | With Blinding Override                      | 26             | 50             | 70             |
| **Payment ReceiveTlvs**                                                                                                    |
|                     | For Offer                                   | 128            | 150            | 140            |
|                     | For Refund                                  | 53             | 100            | 70             |

Reply:
So, just looking at messages, with P=70 we can differentiate the two different types of outbound payments (i.e., for an offer or refund). So P=50 is better when considering that, IIUC. But with P=50, the invoice request case is detectable from the others.

Reply:
Yep, I see that being an issue too.
Another thought—what if we padded to the size of the largest packet, say 128 bytes? That way, all the packets would be uniform in size, at least for standard cases (excluding custom ones). But I do wonder if that might be overkill, especially for much smaller packets, like those around 27-28 bytes. What do you think?

Reply:
sats or msats?

Reply:
That seems rather larger for the minimum. Are you considering the largest we'd see rather than a typical value?

Reply:
Hello!

I’ve updated the `htlc_minimum_msat` to 1 billion (1 million sats) and adjusted `quantity` to `None` (instead of `Some(1)`) to better align with expected values.

Here’s the branch link for you to check out: [branch](https://github.com/shaavan/rust-lightning/tree/padding-test).

I also ran an analysis to determine the best `P` value with minimal added padding for the payment TLVs, especially focusing on `Payment::ReceiveTlvs`. Here’s a quick look at the results:

| Type                     | Property                | Original Sizes | Rounded (P=29)         | Rounded (P=30)         | Rounded (P=32)          |
|--------------------------|-------------------------|----------------|-------------------------|-------------------------|--------------------------|
| **Payment ForwardTlvs**  |                         |                |                         |                         |                          |
|                          | With no Blinding Override | 29             | 29 x 1 = 29 (No Padding) | 30 x 1 = 30 (+1 size)   | 32 x 1 = 32 (+3 size)    |
|                          | With Blinding Override    | 29             | 29 x 1 = 29 (No Padding) | 30 x 1 = 30 (+1 size)   | 32 x 1 = 32 (+3 size)    |
| **Payment ReceiveTlvs**  |                         |                |                         |                         |                          |
|                          | For Offer               | 128            | 29 x 5 = 145 (+17 size) | 30 x 5 = 150 (+22 size) | 32 x 4 = 128 (No Padding) |
|                          | For Refund              | 56             | 29 x 2 = 58 (+ 2 size)  | 30 x 2 = 60 (+4 size)   | 32 x 2 = 64 (+8 size)    |

From the looks of it, `P=32` seems to be the optimal choice for our case with payment TLVs.
Let me know what you think about it! Thanks a lot!

Reply:
Wouldn't this depend on how many forward TLVs are included?

Reply:
Hi!

So, I ran an analysis for different numbers of `ForwardTlvs`. When there are only a few `ForwardTlvs`, the best padding value, P, switches between `Offer` and `Refund`. However, once the count goes over 5, `P=29` generally becomes the optimal choice.

Here’s the table I put together for reference. (I removed the P=30 column since it wasn’t optimal in any scenario.)

| ForwardTlvs Count | ReceiveTlvs Type | Total Size (P=29) | Total Size (P=32) |
| --- | --- | --- | --- |
| 1 | Offer | 174 | *160 |
| 1 | Refund | *87 | 96 |
| 2 | Offer | 203 | *192 |
| 2 | Refund | *116 | 128 |
| 3 | Offer | 232 | *224 |
| 3 | Refund | *145 | 160 |
| 4 | Offer | 261 | *256 |
| 4 | Refund | *174 | 192 |
| 5 | Offer | 290 | *288 |
| 5 | Refund | *203 | 224 |
| 6 | Offer | *319 | 320 |
| 6 | Refund | *232 | 256 |
| 7 | Offer | *348 | 352 |
| 7 | Refund | *261 | 288 |
| 8 | Offer | *377 | 384 |
| 8 | Refund | *290 | 320 |
| 9 | Offer | *406 | 416 |
| 9 | Refund | *319 | 352 |
| 10 | Offer | *435 | 448 |
| 10 | Refund | *348 | 384 |

By the way, in practice, what's the general number of forwardTlvs present in a blinded path?
Thanks a lot!

Reply:
> So, I ran an analysis for different numbers of `ForwardTlvs`. When there are only a few `ForwardTlvs`, the best padding value, P, switches between `Offer` and `Refund`. However, once the count goes over 5, `P=29` generally becomes the optimal choice.

Thanks for putting this together.!

> By the way, in practice, what's the general number of forwardTlvs present in a blinded path? Thanks a lot!

Probably no more than a two or three. I'd lean towards using 32 or something slightly higher to prevent a quantity in the `ReceiveTlvs` from adding another block of padding. Though including a payer note will already cause that.

Reply:
Update the Payment Tlvs padding to `35 ` [pr3177.26](https://github.com/shaavan/rust-lightning/commits/pr3177.26)
Thanks a lot for the discussion and pointers! ✨

Reply:
We'll need to update the analysis now that #3435 has been merged since `payment::ReceiveTlvs` now contains authentication data. Possibly doesn't change anything.

Reply:
Hi! 

I've updated the [branch](https://github.com/shaavan/rust-lightning/tree/padding-test) with the new analysis.

The sizes of `ReceiveTlvs` for both **Offers** and **Refunds** have significantly changed.  
Due to these changes, `P=29` is now the most optimal padding option for all cases of `ForwardTlvs`.

## **Analysis of Sizes and Added Padding for Forward & ReceiveTlvs**

| Type                     | Property                  | Original Sizes | Rounded (P=29)    | Rounded (P=30)  | Rounded (P=32)  |
|--------------------------|---------------------------|----------------|--------------------|-----------------|-----------------|
| **Payment ForwardTlvs** |                           |                |                    |                 |                 |
|                          | With no Blinding Override | 29             | 29 (No Padding)    | 30 (+1 size)    | 32 (+3 size)    |
|                          | With Blinding Override    | 29             | 29 (No Padding)    | 30 (+1 size)    | 32 (+3 size)    |
| **Payment ReceiveTlvs**  |                           |                |                    |                 |                 |
|                          | For Offer                | 196            | 203 (+7 size)      | 210 (+14 size)  | 224 (+28 size)  |
|                          | For Refund               | 110            | 116 (+6 size)      | 120 (+10 size)  | 128 (+18 size)  |

## **Comparative Analysis of Total Size of Blinded Path for ForwardTlvs (0–10)**

### For Offer:
| Number of Forward TLVs | Total Size (P=29) | Total Size (P=30) | Total Size (P=32) |
|-------------------------|-------------------|-------------------|-------------------|
| 0                       | 203               | 210               | 224               |
| 1                       | 232               | 240               | 256               |
| 2                       | 261               | 270               | 288               |
| 3                       | 290               | 300               | 320               |
| 4                       | 319               | 330               | 352               |
| 5                       | 348               | 360               | 384               |
| 6                       | 377               | 390               | 416               |
| 7                       | 406               | 420               | 448               |
| 8                       | 435               | 450               | 480               |
| 9                       | 464               | 480               | 512               |
| 10                      | 493               | 510               | 544               |

### For Refund:
| Number of Forward TLVs | Total Size (P=29) | Total Size (P=30) | Total Size (P=32) |
|-------------------------|-------------------|-------------------|-------------------|
| 0                       | 116               | 120               | 128               |
| 1                       | 145               | 150               | 160               |
| 2                       | 174               | 180               | 192               |
| 3                       | 203               | 210               | 224               |
| 4                       | 232               | 240               | 256               |
| 5                       | 261               | 270               | 288               |
| 6                       | 290               | 300               | 320               |
| 7                       | 319               | 330               | 352               |
| 8                       | 348               | 360               | 384               |
| 9                       | 377               | 390               | 416               |
| 10                      | 406               | 420               | 448               |

Reply:
Discussed offline. The new analysis included `human_readable_name`, which we should exclude since the length is arbitrary.

Reply:
With the updated analysis, the size of the "For Offer" case has changed to 182. Taking this into account, here’s the revised analysis:

The updated findings indicate that **P=31 consistently results in the smallest net Blinded Path size for the "For Offer" case**, whereas **P=29 minimizes the size of the Blinded Path for the "For Refund" case.**

## **Analysis of Sizes and Added Padding for Forward & ReceiveTlvs**

| Type                  | Property                    | Original Sizes | Rounded (P=29)            | Rounded (P=30)            | Rounded (P=31)            | Rounded (P=32)            |
|-----------------------|-----------------------------|----------------|---------------------------|---------------------------|---------------------------|---------------------------|
| **Payment ForwardTlvs** |                             |                |                           |                           |                           |                           |
|                       | With no Blinding Override   | 29             | 29 x 1 = 29 (No Padding)  | 30 x 1 = 30 (+1 size)     | 31 x 1 = 31 (+2 size)     | 32 x 1 = 32 (+3 size)     |
|                       | With Blinding Override      | 29             | 29 x 1 = 29 (No Padding)  | 30 x 1 = 30 (+1 size)     | 31 x 1 = 31 (+2 size)     | 32 x 1 = 32 (+3 size)     |
| **Payment ReceiveTlvs** |                             |                |                           |                           |                           |                           |
|                       | For Offer                  | 182            | 29 x 7 = 203 (+21 size)   | 30 x 7 = 210 (+28 size)   | 31 x 6 = 186 (+4 size)    | 32 x 6 = 192 (+10 size)   |
|                       | For Refund                 | 110            | 29 x 4 = 116 (+6 size)    | 30 x 4 = 120 (+10 size)   | 31 x 4 = 124 (+14 size)   | 32 x 4 = 128 (+18 size)   |

## **Comparative Analysis of Total Size of Blinded Path for ForwardTlvs (0–10)**

### For Offer

| Number of Forward TLVs | Total Size (P=29) | Total Size (P=30) | Total Size (P=31) | Total Size (P=32) |
|-------------------------|-------------------|-------------------|-------------------|-------------------|
| 0                       | 203               | 210               | 186               | 192               |
| 1                       | 261               | 269               | 246               | 253               |
| 2                       | 319               | 328               | 306               | 314               |
| 3                       | 377               | 357               | 366               | 375               |
| 4                       | 435               | 416               | 426               | 436               |
| 5                       | 493               | 475               | 486               | 497               |
| 6                       | 551               | 534               | 546               | 558               |
| 7                       | 609               | 593               | 606               | 619               |
| 8                       | 667               | 652               | 666               | 680               |
| 9                       | 725               | 711               | 726               | 741               |
| 10                      | 783               | 770               | 786               | 802               |

### For Refund

| Number of Forward TLVs | Total Size (P=29) | Total Size (P=30) | Total Size (P=31) | Total Size (P=32) |
|-------------------------|-------------------|-------------------|-------------------|-------------------|
| 0                       | 116               | 120               | 124               | 128               |
| 1                       | 174               | 179               | 184               | 189               |
| 2                       | 232               | 238               | 244               | 250               |
| 3                       | 290               | 297               | 304               | 311               |
| 4                       | 348               | 356               | 364               | 372               |
| 5                       | 406               | 415               | 424               | 433               |
| 6                       | 464               | 474               | 484               | 494               |
| 7                       | 522               | 533               | 544               | 555               |
| 8                       | 580               | 592               | 604               | 616               |
| 9                       | 638               | 651               | 664               | 677               |
| 10                      | 696               | 710               | 724               | 738               |

Reply:
Something looks off when adding the `ForwardTlvs`. I'd expect each row to be divisible by `P`, but that's not the case for `P > 0`. IIUC, each additional `ForwardTlvs` should add `2 * P`.

Reply:
Hi!  
Thanks so much for the nudge!  

I’ve updated the table in the comment above, and with the corrected data, it looks like `P=30` often performs better for the Offer case, while `P=29` generally edges out `P=30` for the Refund case.  

Curious to hear your thoughts on this!  

Btw, I also put together a Python script to analyze the data with a grayscale visualization to better see which values work better across all cases. You can check it out here: [Gist Link](https://gist.github.com/shaavan/97be8e670dd9cad79152be58e84ed8eb).  

Looking forward to your thought!
Thanks!

Reply:
Introduce specific paddings for `Offer` & `Refund` payee tlvs in [pr3177.28](https://github.com/shaavan/rust-lightning/commits/pr3177.28)
Thanks a lot for all the discussions & pointers!